src/lib.rs

```diff
@@ -738,6 +807,18 @@ mod op {
             (Value::Unknown, Value::Unknown) => Value::Unknown,
             (Value::Union(a), b) | (b, Value::Union(a)) => union_bin_op(&a, b, eq),
             (Value::Nil, Value::Nil) => Value::Bool(true),
+            (Value::List(a), Value::List(b))  => {
+                let a = a.borrow();
+                let b = b.borrow();
+                if a.len() != b.len() {
+                    return Value::Bool(false);
+                }
+                a.iter().zip(b.iter()).find_map(
+                    |(a, b)| match eq(a, b) {
+                        Value::Bool(true) => None,
+                        a => Some(a),
+                    }).unwrap_or(Value::Bool(true))
```

Comment:
```suggestion
                Value::Bool(a.iter().zip(b.iter()).all(|(a, b)| matches!(eq(a, b), Value::Bool(true))))
```
This could be made a bit shorter if we made a `Value::try_inner` that's generic over the inner primitives.

Similar changes can be made for the other ops here.

Reply:
Also, tuples can be written the same way.

Reply:
Something like this
```diff
diff --git a/src/lib.rs b/src/lib.rs
index 4a78ecc..349c315 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -268,6 +268,12 @@ impl Type {
     }
 }
 
+trait ToInner<T> {
+    fn to_inner(&self) -> Option<T>;
+
+    unsafe fn to_inner_unchecked(&self) -> T;
+}
+
 #[derive(Clone)]
 pub enum Value {
     Ty(Type),
@@ -289,6 +295,19 @@ pub enum Value {
     Nil,
 }
 
+impl ToInner<bool> for Value {
+    fn to_inner(&self) -> Option<bool> {
+        match self {
+            Value::Bool(b) => Some(*b),
+            _ => None,
+        }
+    }
+
+    unsafe fn to_inner_unchecked(&self) -> bool {
+        self.to_inner().unwrap_or_else(|| std::hint::unreachable_unchecked())
+    }
+}
+
 impl Debug for Value {
     fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
@@ -692,6 +711,7 @@ mod op {
     use super::{Type, Value};
     use std::rc::Rc;
     use std::collections::HashSet;
+    use crate::ToInner;
 
     fn tuple_bin_op(a: &Rc<Vec<Value>>, b: &Rc<Vec<Value>>, f: fn (&Value, &Value) -> Value) -> Value {
         Value::Tuple(Rc::new(a.iter().zip(b.iter()).map(|(a, b)| f(a, b)).collect()))
@@ -809,11 +829,7 @@ mod op {
                 if a.len() != b.len() {
                     return Value::Bool(false);
                 }
-                a.iter().zip(b.iter()).find_map(
-                    |(a, b)| match eq(a, b) {
-                        Value::Bool(true) => None,
-                        a => Some(a),
-                    }).unwrap_or(Value::Bool(true))
+                Value::Bool(a.iter().zip(b.iter()).all(|(a, b)| eq(a, b).to_inner().unwrap()))
             }
             _ => Value::Nil,
         }
```
I also made a `_unchecked` since this is to be used in the VM.

Reply:
This is not semantically equivalent. We can return Value::Nil from the comparison which means the comparison doesn't work because of the types. The type semantics have been changed here.